src/Compilers/CSharp/Test/Semantic/Utilities/ValueSetTests.cs

```diff
@@ -0,0 +1,663 @@
+
+        [Fact]
+        public void TestIntersect_02()
+        {
+            for (int i = 0; i < 100; i++)
+            {
+                int i1 = Random.Next(int.MinValue + 1, int.MaxValue);
+                int i2 = Random.Next(int.MinValue, int.MaxValue);
+                if (i1 < i2) (i1, i2) = (i2, i1);
+                if (i1 == i2) continue;
+                IValueSet<int> values1 = ForInt.Related(GreaterThanOrEqual, i1).Intersect(ForInt.Related(LessThanOrEqual, i2));
+                Assert.Equal($"", values1.ToString());
+                IValueSet<int> values2 = ForInt.Related(LessThanOrEqual, i2).Intersect(ForInt.Related(GreaterThanOrEqual, i1));
+                Assert.Equal(values1, values2);
+            }
+        }
+
+        [Fact]
+        public void TestUnion_01()
+        {
+            for (int i = 0; i < 100; i++)
+            {
+                int i1 = Random.Next(int.MinValue + 1, int.MaxValue);
+                int i2 = Random.Next(int.MinValue, int.MaxValue);
+                if (i1 > i2) (i1, i2) = (i2, i1);
+                if ((i1 + 1) >= i2) continue;
+                IValueSet<int> values1 = ForInt.Related(LessThanOrEqual, i1).Union(ForInt.Related(GreaterThanOrEqual, i2));
+                Assert.Equal($"[{int.MinValue}..{i1}],[{i2}..{int.MaxValue}]", values1.ToString());
+                IValueSet<int> values2 = ForInt.Related(GreaterThanOrEqual, i2).Union(ForInt.Related(LessThanOrEqual, i1));
+                Assert.Equal(values1, values2);
+            }
+        }
+
+        [Fact]
+        public void TestUnion_02()
+        {
+            for (int i = 0; i < 100; i++)
+            {
+                int i1 = Random.Next(int.MinValue + 1, int.MaxValue);
+                int i2 = Random.Next(int.MinValue, int.MaxValue);
+                if (i1 < i2) (i1, i2) = (i2, i1);
+                IValueSet<int> values1 = ForInt.Related(LessThanOrEqual, i1).Union(ForInt.Related(GreaterThanOrEqual, i2));
+                Assert.Equal($"[{int.MinValue}..{int.MaxValue}]", values1.ToString());
+                IValueSet<int> values2 = ForInt.Related(GreaterThanOrEqual, i2).Union(ForInt.Related(LessThanOrEqual, i1));
+                Assert.Equal(values1, values2);
+            }
+        }
+
+        [Fact]
+        public void TestComplement_01()
+        {
+            for (int i = 0; i < 100; i++)
+            {
+                int i1 = Random.Next(int.MinValue + 1, int.MaxValue);
+                int i2 = Random.Next(int.MinValue, int.MaxValue);
+                if (i1 > i2) (i1, i2) = (i2, i1);
+                if ((i1 + 1) >= i2) continue;
+                IValueSet<int> values1 = ForInt.Related(LessThanOrEqual, i1).Union(ForInt.Related(GreaterThanOrEqual, i2));
+                Assert.Equal($"[{int.MinValue}..{i1}],[{i2}..{int.MaxValue}]", values1.ToString());
+                IValueSet<int> values2 = values1.Complement();
+                Assert.Equal($"[{i1 + 1}..{i2 - 1}]", values2.ToString());
+            }
```

Comment:
>} [](start = 12, length = 1)

Consider testing `values2.Complement() == values1`.

Reply:
In the next PR.

---
In reply to: [383411399](https://github.com/dotnet/roslyn/pull/41503#discussion_r383411399) [](ancestors = 383411399)